File: TCPIP/src/17/tcp_state_manager.cpp
#include <iostream>
#include <string>
#include <memory>
#include <chrono>
#include <thread>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netinet/tcp.h>
#include <arpa/inet.h>
#include <unistd.h>
#include <fcntl.h>
#include <cstring>

enum class TCPState {
    CLOSED,
    LISTEN,
    SYN_SENT,
    SYN_RECEIVED,
    ESTABLISHED,
    FIN_WAIT1,
    FIN_WAIT2,
    CLOSE_WAIT,
    CLOSING,
    LAST_ACK,
    TIME_WAIT,
    UNKNOWN
};

class TCPConnection {
private:
    int sockfd_;
    TCPState current_state_;
    std::string remote_addr_;
    uint16_t remote_port_;
    
    TCPState map_kernel_state(int kernel_state) {
        switch(kernel_state) {
            case TCP_ESTABLISHED: return TCPState::ESTABLISHED;
            case TCP_SYN_SENT: return TCPState::SYN_SENT;
            case TCP_SYN_RECV: return TCPState::SYN_RECEIVED;
            case TCP_FIN_WAIT1: return TCPState::FIN_WAIT1;
            case TCP_FIN_WAIT2: return TCPState::FIN_WAIT2;
            case TCP_TIME_WAIT: return TCPState::TIME_WAIT;
            case TCP_CLOSE: return TCPState::CLOSED;
            case TCP_CLOSE_WAIT: return TCPState::CLOSE_WAIT;
            case TCP_LAST_ACK: return TCPState::LAST_ACK;
            case TCP_LISTEN: return TCPState::LISTEN;
            case TCP_CLOSING: return TCPState::CLOSING;
            default: return TCPState::UNKNOWN;
        }
    }
    
    std::string state_to_string(TCPState state) const {
        switch(state) {
            case TCPState::CLOSED: return "CLOSED";
            case TCPState::LISTEN: return "LISTEN";
            case TCPState::SYN_SENT: return "SYN_SENT";
            case TCPState::SYN_RECEIVED: return "SYN_RECEIVED";
            case TCPState::ESTABLISHED: return "ESTABLISHED";
            case TCPState::FIN_WAIT1: return "FIN_WAIT1";
            case TCPState::FIN_WAIT2: return "FIN_WAIT2";
            case TCPState::CLOSE_WAIT: return "CLOSE_WAIT";
            case TCPState::CLOSING: return "CLOSING";
            case TCPState::LAST_ACK: return "LAST_ACK";
            case TCPState::TIME_WAIT: return "TIME_WAIT";
            default: return "UNKNOWN";
        }
    }
    
public:
    TCPConnection() : sockfd_(-1), current_state_(TCPState::CLOSED), remote_port_(0) {}
    
    ~TCPConnection() {
        if (sockfd_ >= 0) {
            close(sockfd_);
        }
    }
    
    bool create_socket() {
        sockfd_ = socket(AF_INET, SOCK_STREAM, 0);
        if (sockfd_ < 0) {
            std::cerr << "Failed to create socket: " << strerror(errno) << std::endl;
            return false;
        }
        current_state_ = TCPState::CLOSED;
        std::cout << "Socket created, state: " << state_to_string(current_state_) << std::endl;
        return true;
    }
    
    TCPState get_current_state() {
        if (sockfd_ < 0) return TCPState::CLOSED;
        
        struct tcp_info info;
        socklen_t info_len = sizeof(info);
        
        if (getsockopt(sockfd_, IPPROTO_TCP, TCP_INFO, &info, &info_len) == 0) {
            current_state_ = map_kernel_state(info.tcpi_state);
        }
        return current_state_;
    }
    
    void print_state_transition(const std::string& event) {
        TCPState old_state = current_state_;
        TCPState new_state = get_current_state();
        
        if (old_state != new_state) {
            std::cout << "State transition [" << event << "]: " 
                      << state_to_string(old_state) << " -> " 
                      << state_to_string(new_state) << std::endl;
        } else {
            std::cout << "Current state [" << event << "]: " 
                      << state_to_string(new_state) << std::endl;
        }
    }
    
    bool connect(const std::string& host, uint16_t port) {
        remote_addr_ = host;
        remote_port_ = port;
        
        struct sockaddr_in server_addr;
        std::memset(&server_addr, 0, sizeof(server_addr));
        server_addr.sin_family = AF_INET;
        server_addr.sin_port = htons(port);
        inet_pton(AF_INET, host.c_str(), &server_addr.sin_addr);
        
        std::cout << "\nInitiating connection to " << host << ":" << port << std::endl;
        
        // Make non-blocking to observe SYN_SENT
        int flags = fcntl(sockfd_, F_GETFL, 0);
        fcntl(sockfd_, F_SETFL, flags | O_NONBLOCK);
        
        int result = ::connect(sockfd_, (struct sockaddr*)&server_addr, sizeof(server_addr));
        
        if (result < 0 && errno == EINPROGRESS) {
            print_state_transition("connect() called");
            
            // Wait for connection to complete
            fd_set write_fds;
            struct timeval tv = {5, 0};
            
            FD_ZERO(&write_fds);
            FD_SET(sockfd_, &write_fds);
            
            if (select(sockfd_ + 1, nullptr, &write_fds, nullptr, &tv) > 0) {
                int error;
                socklen_t len = sizeof(error);
                getsockopt(sockfd_, SOL_SOCKET, SO_ERROR, &error, &len);
                
                if (error == 0) {
                    fcntl(sockfd_, F_SETFL, flags); // Restore blocking mode
                    print_state_transition("connection established");
                    return true;
                }
            }
        }
        
        std::cerr << "Connection failed: " << strerror(errno) << std::endl;
        return false;
    }
    
    ssize_t send(const std::string& data) {
        if (current_state_ != TCPState::ESTABLISHED) {
            std::cerr << "Cannot send: not in ESTABLISHED state" << std::endl;
            return -1;
        }
        
        ssize_t sent = ::send(sockfd_, data.c_str(), data.size(), 0);
        std::cout << "Sent " << sent << " bytes" << std::endl;
        return sent;
    }
    
    ssize_t receive(char* buffer, size_t size) {
        ssize_t received = ::recv(sockfd_, buffer, size, 0);
        
        if (received == 0) {
            print_state_transition("received FIN (recv=0)");
        } else if (received > 0) {
            std::cout << "Received " << received << " bytes" << std::endl;
        }
        
        return received;
    }
    
    void shutdown_write() {
        std::cout << "\nInitiating graceful shutdown (write side)" << std::endl;
        ::shutdown(sockfd_, SHUT_WR);
        print_state_transition("shutdown(SHUT_WR)");
    }
    
    void close_connection() {
        if (sockfd_ >= 0) {
            std::cout << "\nClosing connection" << std::endl;
            print_state_transition("before close()");
            ::close(sockfd_);
            sockfd_ = -1;
            current_state_ = TCPState::CLOSED;
            std::cout << "Connection closed, socket in TIME_WAIT or CLOSED" << std::endl;
        }
    }
    
    // Demonstrate half-close scenario
    void demonstrate_half_close() {
        if (current_state_ != TCPState::ESTABLISHED) {
            std::cerr << "Not in ESTABLISHED state" << std::endl;
            return;
        }
        
        std::cout << "\n=== Demonstrating Half-Close ===\n";
        
        // Send final data
        send("Final message before half-close\n");
        
        // Close write side (moves to FIN_WAIT1/FIN_WAIT2)
        shutdown_write();
        
        // Can still receive data
        char buffer[1024];
        std::cout << "Can still receive data in half-closed state..." << std::endl;
        ssize_t n = receive(buffer, sizeof(buffer));
        
        if (n > 0) {
            std::cout << "Received: " << std::string(buffer, n);
        }
        
        // Continue receiving until FIN from remote
        while ((n = receive(buffer, sizeof(buffer))) > 0) {
            std::cout << "Received: " << std::string(buffer, n);
        }
        
        print_state_transition("after receiving remote FIN");
    }
    
    int get_socket() const { return sockfd_; }
};

// Server class to demonstrate passive open states
class TCPServer {
private:
    int listen_fd_;
    TCPState listen_state_;
    uint16_t port_;
    
public:
    TCPServer(uint16_t port) : listen_fd_(-1), listen_state_(TCPState::CLOSED), port_(port) {}
    
    ~TCPServer() {
        if (listen_fd_ >= 0) {
            close(listen_fd_);
        }
    }
    
    bool start() {
        listen_fd_ = socket(AF_INET, SOCK_STREAM, 0);
        if (listen_fd_ < 0) {
            std::cerr << "Failed to create socket" << std::endl;
            return false;
        }
        
        // Enable address reuse
        int reuse = 1;
        setsockopt(listen_fd_, SOL_SOCKET, SO_REUSEADDR, &reuse, sizeof(reuse));
        
        struct sockaddr_in server_addr;
        std::memset(&server_addr, 0, sizeof(server_addr));
        server_addr.sin_family = AF_INET;
        server_addr.sin_addr.s_addr = INADDR_ANY;
        server_addr.sin_port = htons(port_);
        
        if (bind(listen_fd_, (struct sockaddr*)&server_addr, sizeof(server_addr)) < 0) {
            std::cerr << "Bind failed: " << strerror(errno) << std::endl;
            return false;
        }
        
        if (listen(listen_fd_, 5) < 0) {
            std::cerr << "Listen failed" << std::endl;
            return false;
        }
        
        listen_state_ = TCPState::LISTEN;
        std::cout << "Server listening on port " << port_ << ", state: LISTEN" << std::endl;
        return true;
    }
    
    std::unique_ptr<TCPConnection> accept_connection() {
        struct sockaddr_in client_addr;
        socklen_t client_len = sizeof(client_addr);
        
        std::cout << "\nWaiting for connection..." << std::endl;
        
        int conn_fd = accept(listen_fd_, (struct sockaddr*)&client_addr, &client_len);
        if (conn_fd < 0) {
            std::cerr << "Accept failed" << std::endl;
            return nullptr;
        }
        
        char client_ip[INET_ADDRSTRLEN];
        inet_ntop(AF_INET, &client_addr.sin_addr, client_ip, sizeof(client_ip));
        
        std::cout << "Accepted connection from " << client_ip << ":" 
                  << ntohs(client_addr.sin_port) << std::endl;
        
        auto conn = std::make_unique<TCPConnection>();
        // Manually set the accepted socket
        conn->~TCPConnection();
        new (conn.get()) TCPConnection();
        // Note: This is simplified; in production, you'd properly handle this
        
        std::cout << "Connection state: ESTABLISHED" << std::endl;
        return conn;
    }
};

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cout << "Usage:\n";
        std::cout << "  Server: " << argv[0] << " server <port>\n";
        std::cout << "  Client: " << argv[0] << " client <host> <port>\n";
        return 1;
    }
    
    std::string mode = argv[1];
    
    if (mode == "client") {
        std::string host = argc > 2 ? argv[2] : "127.0.0.1";
        uint16_t port = argc > 3 ? std::stoi(argv[3]) : 8080;
        
        std::cout << "=== TCP Client State Demonstration ===\n";
        
        TCPConnection conn;
        if (!conn.create_socket()) return 1;
        
        if (conn.connect(host, port)) {
            // Send data
            conn.send("Hello from C++ client\n");
            
            // Receive response
            char buffer[1024];
            conn.receive(buffer, sizeof(buffer));
            
            // Demonstrate half-close
            conn.demonstrate_half_close();
            
            // Final close
            conn.close_connection();
        }
        
    } else if (mode == "server") {
        uint16_t port = argc > 2 ? std::stoi(argv[2]) : 8080;
        
        std::cout << "=== TCP Server State Demonstration ===\n";
        
        TCPServer server(port);
        if (!server.start()) return 1;
        
        // Accept one connection
        auto conn = server.accept_connection();
        if (conn) {
            std::cout << "Connection handling would go here\n";
            std::this_thread::sleep_for(std::chrono::seconds(2));
        }
    }
    
    return 0;
}

File: TCPIP/src/17/tcp_state_manager.rs
use std::io::{self, Read, Write};
use std::net::{TcpListener, TcpStream, Shutdown};
use std::time::Duration;
use std::thread;

#[derive(Debug, Clone, Copy, PartialEq)]
enum TcpState {
    Closed,
    Listen,
    SynSent,
    SynReceived,
    Established,
    FinWait1,
    FinWait2,
    CloseWait,
    Closing,
    LastAck,
    TimeWait,
}

impl TcpState {
    fn as_str(&self) -> &'static str {
        match self {
            TcpState::Closed => "CLOSED",
            TcpState::Listen => "LISTEN",
            TcpState::SynSent => "SYN_SENT",
            TcpState::SynReceived => "SYN_RECEIVED",
            TcpState::Established => "ESTABLISHED",
            TcpState::FinWait1 => "FIN_WAIT1",
            TcpState::FinWait2 => "FIN_WAIT2",
            TcpState::CloseWait => "CLOSE_WAIT",
            TcpState::Closing => "CLOSING",
            TcpState::LastAck => "LAST_ACK",
            TcpState::TimeWait => "TIME_WAIT",
        }
    }
}

/// Wrapper around TcpStream that tracks state transitions
struct TcpConnection {
    stream: TcpStream,
    current_state: TcpState,
    peer_addr: String,
}

impl TcpConnection {
    fn new(stream: TcpStream) -> io::Result<Self> {
        let peer_addr = stream.peer_addr()?.to_string();
        
        Ok(TcpConnection {
            stream,
            current_state: TcpState::Established,
            peer_addr,
        })
    }
    
    fn from_connect(addr: &str) -> io::Result<Self> {
        println!("Initiating connection to {}", addr);
        println!("State transition: CLOSED -> SYN_SENT");
        
        // TcpStream::connect handles the three-way handshake
        let stream = TcpStream::connect(addr)?;
        
        println!("State transition: SYN_SENT -> ESTABLISHED");
        
        let peer_addr = stream.peer_addr()?.to_string();
        
        Ok(TcpConnection {
            stream,
            current_state: TcpState::Established,
            peer_addr,
        })
    }
    
    fn print_state(&self, event: &str) {
        println!("[{}] Current state: {} (peer: {})", 
                 event, self.current_state.as_str(), self.peer_addr);
    }
    
    fn state_transition(&mut self, event: &str, new_state: TcpState) {
        if self.current_state != new_state {
            println!("State transition [{}]: {} -> {}", 
                     event, self.current_state.as_str(), new_state.as_str());
            self.current_state = new_state;
        }
    }
    
    fn send_data(&mut self, data: &[u8]) -> io::Result<usize> {
        if self.current_state != TcpState::Established {
            return Err(io::Error::new(
                io::ErrorKind::NotConnected,
                format!("Cannot send in {} state", self.current_state.as_str())
            ));
        }
        
        let sent = self.stream.write(data)?;
        self.stream.flush()?;
        println!("Sent {} bytes", sent);
        Ok(sent)
    }
    
    fn receive_data(&mut self, buffer: &mut [u8]) -> io::Result<usize> {
        match self.stream.read(buffer) {
            Ok(0) => {
                // Received FIN from remote
                self.state_transition("received FIN", TcpState::CloseWait);
                Ok(0)
            }
            Ok(n) => {
                println!("Received {} bytes", n);
                Ok(n)
            }
            Err(e) => Err(e),
        }
    }
    
    fn shutdown_write(&mut self) -> io::Result<()> {
        println!("\nInitiating graceful shutdown (write side)");
        
        // This sends FIN, transitioning to FIN_WAIT1
        self.state_transition("shutdown(Write)", TcpState::FinWait1);
        self.stream.shutdown(Shutdown::Write)?;
        
        // After receiving ACK for our FIN, moves to FIN_WAIT2
        thread::sleep(Duration::from_millis(100));
        self.state_transition("received ACK for FIN", TcpState::FinWait2);
        
        Ok(())
    }
    
    fn close(mut self) {
        println!("\nClosing connection");
        self.print_state("before close");
        
        // Drop causes the connection to close
        // This sends FIN if not already sent
        drop(self.stream);
        
        println!("Connection closed, entering TIME_WAIT state (2*MSL wait)");
    }
    
    /// Demonstrates half-close scenario
    fn demonstrate_half_close(&mut self) -> io::Result<()> {
        println!("\n=== Demonstrating Half-Close ===");
        
        // Send final message
        self.send_data(b"Final message before half-close\n")?;
        
        // Close write side (send FIN)
        self.shutdown_write()?;
        
        // Can still receive data
        println!("Can still receive data in half-closed state...");
        let mut buffer = [0u8; 1024];
        
        loop {
            match self.receive_data(&mut buffer) {
                Ok(0) => {
                    println!("Received FIN from remote, connection fully closed");
                    break;
                }
                Ok(n) => {
                    let data = String::from_utf8_lossy(&buffer[..n]);
                    println!("Received in half-closed state: {}", data);
                }
                Err(e) => {
                    println!("Error receiving: {}", e);
                    break;
                }
            }
        }
        
        Ok(())
    }
}

/// Server that demonstrates passive open states
struct TcpServer {
    listener: TcpListener,
    state: TcpState,
}

impl TcpServer {
    fn new(addr: &str) -> io::Result<Self> {
        println!("Creating server socket");
        println!("State: CLOSED");
        
        let listener = TcpListener::bind(addr)?;
        println!("Bound to {}", addr);
        
        // SO_REUSEADDR is set by default in Rust
        
        println!("State transition: CLOSED -> LISTEN");
        
        Ok(TcpServer {
            listener,
            state: TcpState::Listen,
        })
    }
    
    fn accept(&mut self) -> io::Result<TcpConnection> {
        println!("\nWaiting for connection (state: {})...", self.state.as_str());
        
        let (stream, addr) = self.listener.accept()?;
        
        println!("Accepted connection from {}", addr);
        println!("State transition: LISTEN -> SYN_RECEIVED -> ESTABLISHED");
        
        TcpConnection::new(stream)
    }
}

/// Client demonstration
fn run_client(server_addr: &str) -> io::Result<()> {
    println!("=== TCP Client State Demonstration ===\n");
    
    // Connect (CLOSED -> SYN_SENT -> ESTABLISHED)
    let mut conn = TcpConnection::from_connect(server_addr)?;
    conn.print_state("after connect");
    
    // Set timeouts
    conn.stream.set_read_timeout(Some(Duration::from_secs(5)))?;
    conn.stream.set_write_timeout(Some(Duration::from_secs(5)))?;
    
    // Send some data
    conn.send_data(b"Hello from Rust client\n")?;
    
    // Receive response
    let mut buffer = [0u8; 1024];
    if let Ok(n) = conn.receive_data(&mut buffer) {
        if n > 0 {
            let response = String::from_utf8_lossy(&buffer[..n]);
            println!("Received response: {}", response);
        }
    }
    
    // Demonstrate half-close
    conn.demonstrate_half_close()?;
    
    // Full close
    conn.close();
    
    println!("\nClient finished, socket in TIME_WAIT/CLOSED");
    
    Ok(())
}

/// Server demonstration
fn run_server(bind_addr: &str) -> io::Result<()> {
    println!("=== TCP Server State Demonstration ===\n");
    
    let mut server = TcpServer::new(bind_addr)?;
    
    println!("Server ready on {}", bind_addr);
    
    // Accept one connection
    let mut conn = server.accept()?;
    conn.print_state("after accept");
    
    // Receive data
    let mut buffer = [0u8; 1024];
    
    loop {
        match conn.receive_data(&mut buffer) {
            Ok(0) => {
                // Client closed (received FIN)
                println!("\nClient initiated close");
                conn.print_state("received FIN");
                
                // Server should close too to avoid staying in CLOSE_WAIT
                println!("Server closing to transition from CLOSE_WAIT");
                
                // Send any final data
                conn.send_data(b"Goodbye from server\n")?;
                
                // Close connection (sends FIN, moves to LAST_ACK)
                println!("State: CLOSE_WAIT -> LAST_ACK -> CLOSED");
                drop(conn);
                
                break;
            }
            Ok(n) => {
                let data = String::from_utf8_lossy(&buffer[..n]);
                println!("Received: {}", data);
                
                // Echo back
                conn.send_data(&buffer[..n])?;
            }
            Err(e) => {
                println!("Error: {}", e);
                break;
            }
        }
    }
    
    println!("\nServer finished");
    
    Ok(())
}

/// Demonstrates TIME_WAIT state and SO_REUSEADDR
fn demonstrate_time_wait() -> io::Result<()> {
    println!("\n=== Demonstrating TIME_WAIT State ===\n");
    
    let addr = "127.0.0.1:9999";
    
    println!("Creating first server on {}", addr);
    {
        let _server = TcpServer::new(addr)?;
        println!("Server in LISTEN state");
        // Server drops here, goes to CLOSED
    }
    
    println!("\nServer closed");
    println!("Attempting to immediately rebind to same address...");
    
    // This works because Rust sets SO_REUSEADDR by default
    match TcpServer::new(addr) {
        Ok(_) => println!("SUCCESS: Rebound immediately (SO_REUSEADDR enabled)"),
        Err(e) => println!("FAILED: {} (would need SO_REUSEADDR)", e),
    }
    
    Ok(())
}

fn main() -> io::Result<()> {
    let args: Vec<String> = std::env::args().collect();
    
    if args.len() < 2 {
        println!("Usage:");
        println!("  Server: {} server [addr:port]", args[0]);
        println!("  Client: {} client [addr:port]", args[0]);
        println!("  TIME_WAIT demo: {} timewait", args[0]);
        return Ok(());
    }
    
    match args[1].as_str() {
        "server" => {
            let addr = if args.len() > 2 { &args[2] } else { "127.0.0.1:8080" };
            run_server(addr)?;
        }
        "client" => {
            let addr = if args.len() > 2 { &args[2] } else { "127.0.0.1:8080" };
            run_client(addr)?;
        }
        "timewait" => {
            demonstrate_time_wait()?;
        }
        _ => {
            println!("Unknown command: {}", args[1]);
        }
    }
    
    Ok(())
}